frame/transaction-storage/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+
+/// State data for a stored transaction.
+#[derive(Encode, Decode, sp_runtime::RuntimeDebug, PartialEq, Eq)]
+pub struct TransactionInfo {
+	/// Chunk trie root.
+	chunk_root: <BlakeTwo256 as Hash>::Output,
+	/// Plain hash of indexed data.
+	content_hash: <BlakeTwo256 as Hash>::Output,
+	/// Size of indexed data in bytes.
+	size: u32,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// A dispatchable call.
+		type Call: Parameter + Dispatchable<Origin=Self::Origin> + GetDispatchInfo + From<frame_system::Call<Self>>;
+		/// The currency trait.
+		type Currency: ReservableCurrency<Self::AccountId>;
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Insufficient account balance.
+		InsufficientFunds,
+		/// Invalid configuration.
+		NotConfigured,
+		/// Renewed extrinsic is not found.
+		RenewedNotFound,
+		/// Attempting to store empty transaction
+		EmptyTransaction,
+		/// Proof was not expected in this block.
+		UnexpectedProof,
+		/// Proof failed verification.
+		InvalidProof,
+		/// Missing storage proof.
+		MissingProof,
+		/// Unable to verify proof becasue state data is missing.
+		MissingStateData,
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			let period = <StoragePeriod<T>>::get();
+			let obsolete = n.saturating_sub(period);
+			let transaction_count = if obsolete > Zero::zero() {
+				<TransactionCount<T>>::get(obsolete)
+			} else {
+				0
+			};
+			// return weight for `on_finalize` does
+			T::WeightInfo::on_finalize(transaction_count)
+		}
+
+		fn on_finalize(n: T::BlockNumber) {
+			<TransactionCount<T>>::insert(n, <Counter<T>>::get());
+			<Counter<T>>::kill();
+			// Drop obsolete roots
+			let period = <StoragePeriod<T>>::get();
+			let obsolete = n.saturating_sub(period);
+			if obsolete > Zero::zero() {
+				<TransactionRoots<T>>::remove_prefix(obsolete);
+				<TransactionCount<T>>::remove(obsolete);
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Index and store data on chain. Minimum data size is 1 bytes, maximum is `MAX_DATA_SIZE`.
+		/// Data will be removed after `STORAGE_PERIOD` blocks, unless `renew` is called.
+		/// # <weight>
+		/// - n*log(n) of data size, as all data is pushed to an in-memory trie.
+		/// Additionally contains a DB write.
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::store(data.len() as u32))]
+		pub(super) fn store(
+			origin: OriginFor<T>,
+			data: Vec<u8>,
```

Comment:
can use `BoundedVec` which will be type safe for the length

Reply:
Opted to make it runtime configurable, rather then compile time.